mattrib.1
'\" t
.TH mattrib 1 "02Dec18" mtools-4.0.22
.SH Name
mattrib - change MSDOS file attribute flags
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note\ of\ warning
This manpage has been automatically generated from mtools's texinfo
documentation, and may not be entirely accurate or complete.  See the
end of this man page for details.
.PP
.SH Description
.PP
\&\fR\&\f(CWMattrib\fR is used to change MS-DOS file attribute flags. It has the
following syntax:
.PP
\&\fR\&\f(CWmattrib\fR [\fR\&\f(CW-a|+a\fR] [\fR\&\f(CW-h|+h\fR] [\fR\&\f(CW-r|+r\fR]
[\fR\&\f(CW-s|+s\fR] [\fR\&\f(CW-/\fR]  [\fR\&\f(CW-p\fR] [\fR\&\f(CW-X\fR] \fImsdosfile\fR [ \fImsdosfiles\fR \&... ]
.PP
\&\fR\&\f(CWMattrib\fR adds attribute flags to an MS-DOS file (with the
`\fR\&\f(CW+\fR' operator) or remove attribute flags (with the `\fR\&\f(CW-\fR'
operator).
.PP
\&\fR\&\f(CWMattrib\fR supports the following attribute bits:
.TP
\&\fR\&\f(CWa\fR\ 
Archive bit.  Used by some backup programs to indicate a new file.
.TP
\&\fR\&\f(CWr\fR\ 
Read-only bit.  Used to indicate a read-only file.  Files with this bit
set cannot be erased by \fR\&\f(CWDEL\fR nor modified.
.TP
\&\fR\&\f(CWs\fR\ 
System bit.  Used by MS-DOS to indicate a operating system file.
.TP
\&\fR\&\f(CWh\fR\ 
Hidden bit.  Used to make files hidden from \fR\&\f(CWDIR\fR.
.PP
\&\fR\&\f(CWMattrib\fR supports the following command line flags:
.TP
\&\fR\&\f(CW/\fR\ 
Recursive.  Recursively list the attributes of the files in the subdirectories.
.TP
\&\fR\&\f(CWX\fR\ 
Concise. Prints the attributes without any whitespace padding.  If
neither the "/" option is given, nor the \fImsdosfile\fR contains a
wildcard, and there is only one MS-DOS file parameter on the command
line, only the attribute is printed, and not the filename.  This option
is convenient for scripts
.TP
\&\fR\&\f(CWp\fR\ 
Replay mode.  Outputs a series of mformat commands that will reproduce
the current situation, starting from a situation as left by untarring
the MS-DOS file system.  Commands are only output for attribute settings
that differ from the default (archive bit set for files, unset for
directories).  This option is intended to be used in addition to
tar. The \fR\&\f(CWreadonly\fR attribute is not taken into account, as tar can
set that one itself.
.PP
.SH See\ Also
Mtools' texinfo doc
.SH Viewing\ the\ texi\ doc
This manpage has been automatically generated from mtools's texinfo
documentation. However, this process is only approximative, and some
items, such as crossreferences, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, not all information has
been translated into the manpage version.  Thus I strongly advise you to
use the original texinfo doc.  See the end of this manpage for
instructions how to view the texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips mtools.dvi
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.TP
* \ \ 
To generate a html copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.PP
 
\&\fRA premade html can be found at
\&\fR\&\f(CW\(ifhttp://www.gnu.org/software/mtools/manual/mtools.html\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as html.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.PP


NEWS
v4_0_22
	- Fixed -f flag for mformat (size is KBytes, rather than sectors)
	- Fixed toupper/tolower usage (unsigned char rather than plain signed)
v4_0_21
	- Fixed compilation for MingW
	- After MingW compilation, make sure executable has .exe extension
	- Addressed compiler warnings
	- Fixed length handling in character set conversion (Unicode file names)
	- Fixed matching of character range, when containing Unicode
	characters (mdir "c:test[α-ω].exe")
	- Fixed initialization of my_scsi_cmd constructor
v4_0_20
	- initialize directory entries to 0
	- bad message "Too few sectors" replaced with "Too many sectors"
	- apostrophe in mlabel no longer causes generation of long entry
	- option to fake system date for file creation using the
	SOURCE_DATE_EPOCH environment variables

	- can now be compiled with "clang" compiler
	- fallback function for strndup, for those platforms that don't have it
	- fixed a number of -Wextra warnings

	- new compressed archive formats for uz/lz

	- allow to specify number of reserved sectors for FAT32.
	- file/device locking with timeout (rather than immediate failure)
	- fixed support for BPB-less legacy formats.
	- removed check that disk must be an integer number of tracks.

	- removed .eh/.oh macros from manual pages
v4_0_19
	- Fix for short file names starting with character 0xE5
	(by remapping it to 0x5)
	- mpartition: Partition types closer to what Microsoft uses
	- mformat: figure out LBA geometry as last resort if geometry
	is neither specified in config and/or commandline, nor can be
	queried from the device
	- mformat: use same default cluster size by size as Microsoft
	for FAT32
	- additional sanity checks
	- document how cluster size is picked in mformat.c man page
	- document how partition types are picked in mpartition.c man page
v4_0_18	
	Fix for names of iconv encodings on AIX
	Fix mt_size_t on NetBSD
	Fixed compilation on Mingw
	Fixed doc (especially mformat)
	Fix mformat'ing of FAT12 filesystems with huge cluster sizes
	Minfo prints image file name in mformat command line if an image
	file name was given
	Always generate gzip-compressed RPMs, in order to remain
	compatible with older distributions
	Fixed buffer overflow with drive letter in mclasserase
v4_0_17
	mbadblocks now takes a list of bad blocks (either as sectors
	or as clusters)
	mbadblocks now is able to do write scanning for bad blocks
	mshowfat can show cluster of specific offset
	Enable mtools to deal with very small sector sizes...
	Fixed encoding of all-lowercase names (no need to mangle
	these)
	Consider every directory entry after an ENDMARK (0x00) to be deleted
	After writing a new entry at end of a directory, be sure to also add
	an ENDMARK (0x00)

	Deal with possibility of a NULL pointer being returned by
	localtime during timestamp conversion
v4_0_16
	configure.in fixes
	fixed formatting of fat_size_calculation.tex document
	compatibility with current autoconfig versions
	Make it clear that label is limited to 11 characters
	Fixed typo in initialization of FAT32 info sector
v4_0_15
	Added missing -i option to mshortname
	Splitted .deb package into mtools and floppyd in order to
	match Ubuntu
v4_0_14
	New mshortname command
	Fix floppyd for disks bigger than 2 Gig
	Remove obsolete -z flag
	Remove now unsupported AC_USE_SYSTEM_EXTENSIONS
	Fixed output formatting of mdir if MTOOLS_DOTTED_DIR is set
	Mformat now correctly writes backup boot sector
	Fixed signedness of serial number in mlabel
	Fixed buffer size problem in mlabel
	Make mlabel write backup boot sector if FAT32
	Catch situation where both clear and new label are given to mlabel
	Quote filename parameters to scripts
	Mformat: Close file descriptor for boot sector
	Added lzip support to scripts/uz
	Added Tot_sectors option to mformat
	Fixed hidden sector handling in mformat
	Minfo generates mformat command lines containing new -T option
	Mlabel prints error if label too long
v4_0_13
	Merged Debian patches
v4_0_12
	Mingw compatibility fixes
v4_0_11
	Fixed compiler warnings in mlabel.c and elsewhere
	Fixed h flag in mattrib.c
	Added missing error checking in floppyd and elsewhere
	
v4_0_10
	More copyright stuff...
	Fixed issues with max filesize (was 2GB instead of 4GB, and
	warned only after copying the beginning)
v4_0_9
	More copyright stuff
v4_0_8
	Corrected copyright attributions in the various files
v4_0_7
	Fixed conversion to native on OS/2
	Fix parsing of --help flag
v4_0_6
	Fallback for missing wchar_t iconv codepage on OS/2
	Fixes for LSEEK64 support
	Support for --help that returns a 0 exit status
v4_0_5
	Make setpgrp() usage in floppyd conditional
	Re-instate PACKED around structure (ARM)
	LSEEK64
	
v4_0_4
	BSD support: SCSI, use getuserid/getgroupid in floppyd
	Another attempt at putwc fix for OS/2
	Further GNU fixes
	Fallback for putwc if there is wchar (OS/2)
v4_0_3
	Fix multipart pathname parsing bug in vfat.c (forgot limited length)
	Supplied fallback define for putwc
	Copyright notices in all sources
v4_0_2
	Off-by-2 error in unix_name in file_name.c
v4_0_1
	Missing functions on Solaris
v4_0_0
	Offset for -i-specified image files
v4_0_0_pre2
	Use transliteration to represent characters which don't exist in
	target set
v4_0_0_pre1
	Mtools-4 with Unicode support
	Released 4.0.0_pre1
v20071226
	Debian build files
	Fixed security issue with doctored file names
	64 bit compilation fixes
v20070601
	Fixed misc blunders...
v20070531
	Fixed lots of minor items raised by gcc4
	Merged some of the BSD patches
	New version of amuFormat.sh written in sh rather than csh
	Support for config parameters after -i file
	Released 3.9.11
v20070411
	Added sizecode printing on minfo
	In mformat manpage, use same flag for sector number than printed in
	minfo
	Limit sizecode to 6, else it will overflow max sector size defined in
	msdos.h
v20070308
	Applied mingw patch by Jamey Sharp and Josh Triplett
v20070306
	Fixed doc about /etc/default
v20070305
	Fixed mlabel on read-only disks
v20060626
	Merged Redhat/Fedora patches

mattrib.c
/*  Copyright 1986-1992 Emmet P. Gray.
 *  Copyright 1996-1998,2000-2002,2007,2009 Alain Knaff.
 *  This file is part of mtools.
 *
 *  Mtools is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Mtools is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Mtools.  If not, see <http://www.gnu.org/licenses/>.
 *
 * mattrib.c
 * Change MSDOS file attribute flags
 */

#include "sysincludes.h"
#include "msdos.h"
#include "mtools.h"
#include "mainloop.h"

typedef struct Arg_t {
	char add;
	unsigned char remove;
	struct MainParam_t mp;
	int recursive;
	int doPrintName;
} Arg_t;

static int attrib_file(direntry_t *entry, MainParam_t *mp)
{
	Arg_t *arg=(Arg_t *) mp->arg;

	if(entry->entry != -3) {
		/* if not root directory, change it */
		entry->dir.attr = (entry->dir.attr & arg->remove) | arg->add;
		dir_write(entry);
	}
	return GOT_ONE;
}

static int replay_attrib(direntry_t *entry, MainParam_t *mp UNUSEDP)
{
	if ( (IS_ARCHIVE(entry) && IS_DIR(entry)) ||
		 (!IS_ARCHIVE(entry) && !IS_DIR(entry)) ||
		 IS_SYSTEM(entry) || IS_HIDDEN(entry)) {

		printf("mattrib ");

		if (IS_ARCHIVE(entry) && IS_DIR(entry)) {
			printf("+a ");
		}

		if (!IS_ARCHIVE(entry) && !IS_DIR(entry)) {
			printf("-a ");
		}

		if (IS_SYSTEM(entry)) {
			printf("+s ");
		}

		if (IS_HIDDEN(entry)) {
			printf("+h ");
		}

		fprintPwd(stdout, entry, 1);
		printf("\n");
	}
	return GOT_ONE;
}



static int view_attrib(direntry_t *entry, MainParam_t *mp UNUSEDP)
{
	printf("  ");
	if(IS_ARCHIVE(entry))
		putchar('A');
	else
		putchar(' ');
	fputs("  ",stdout);
	if(IS_SYSTEM(entry))
		putchar('S');
	else
		putchar(' ');
	if(IS_HIDDEN(entry))
		putchar('H');
	else
		putchar(' ');
	if(IS_READONLY(entry))
		putchar('R');
	else
		putchar(' ');
	printf("     ");
	fprintPwd(stdout, entry, 0);
	printf("\n");
	return GOT_ONE;
}


static int concise_view_attrib(direntry_t *entry, MainParam_t *mp)
{
	Arg_t *arg=(Arg_t *) mp->arg;

	if(IS_ARCHIVE(entry))
		putchar('A');
	if(IS_DIR(entry))
		putchar('D');	
	if(IS_SYSTEM(entry))
		putchar('S');
	if(IS_HIDDEN(entry))
		putchar('H');
	if(IS_READONLY(entry))
		putchar('R');
	if(arg->doPrintName) {
		putchar(' ');
		fprintPwd(stdout, entry, 0);
	}
	putchar('\n');
	return GOT_ONE;
}

static int recursive_attrib(direntry_t *entry, MainParam_t *mp)
{
	mp->callback(entry, mp);
	return mp->loop(mp->File, mp, "*");
}


static void usage(int ret) NORETURN;
static void usage(int ret)
{
	fprintf(stderr, "Mtools version %s, dated %s\n", 
		mversion, mdate);
	fprintf(stderr, 
		"Usage: %s [-p] [-a|+a] [-h|+h] [-r|+r] [-s|+s] msdosfile [msdosfiles...]\n",
		progname);
	exit(ret);
}

static int letterToCode(int letter)
{
	switch (toupper(letter)) {
		case 'A':
			return ATTR_ARCHIVE;
		case 'H':
			return ATTR_HIDDEN;
		case 'R':
			return ATTR_READONLY;
		case 'S':
			return ATTR_SYSTEM;
		default:
			usage(1);
	}
}

void mattrib(int argc, char **argv, int type UNUSEDP) NORETURN;
void mattrib(int argc, char **argv, int type UNUSEDP)
{
	Arg_t arg;
	int view;
	int c;
	int concise;
	int replay;
	char *ptr;
	int wantUsage;

	arg.add = 0;
	arg.remove = 0xff;
	arg.recursive = 0;
	arg.doPrintName = 1;
	view = 0;
	concise = 0;
	replay = 0;
	wantUsage = 0;

	if(helpFlag(argc, argv))
		usage(0);
	while ((c = getopt(argc, argv, "i:/ahrsAHRSXp")) != EOF) {
		switch (c) {
			case 'h':
				wantUsage = 1;
				/* FALL THROUGH */
			default:
				arg.remove &= ~letterToCode(c);
				break;
			case 'i':
				set_cmd_line_image(optarg);
				break;
			case 'p':
				replay = 1;
				break;
			case '/':
				arg.recursive = 1;
				break;
			case 'X':
				concise = 1;
				break;
			case '?':
				usage(1);
		}
	}

	if(optind == argc && wantUsage) {
		usage(0);
	}

	for(;optind < argc;optind++) {
		switch(argv[optind][0]) {
			case '+':
				for(ptr = argv[optind] + 1; *ptr; ptr++)
					arg.add |= letterToCode(*ptr);
				continue;
			case '-':
				for(ptr = argv[optind] + 1; *ptr; ptr++)
					arg.remove &= ~letterToCode(*ptr);
				continue;
		}
		break;
	}

	if(arg.remove == 0xff && !arg.add)
		view = 1;

	if (optind >= argc)
		usage(1);

	init_mp(&arg.mp);
	if(view){
		if(concise) {
			arg.mp.callback = concise_view_attrib;
			arg.doPrintName = (argc - optind > 1 ||
					   arg.recursive ||
					   strpbrk(argv[optind], "*[?") != 0);
		} else if (replay) {
			arg.mp.callback = replay_attrib;
		} else
			arg.mp.callback = view_attrib;
		arg.mp.openflags = O_RDONLY;
	} else {
		arg.mp.callback = attrib_file;
		arg.mp.openflags = O_RDWR;
	}

	if(arg.recursive)
		arg.mp.dirCallback = recursive_attrib;

	arg.mp.arg = (void *) &arg;
	arg.mp.lookupflags = ACCEPT_PLAIN | ACCEPT_DIR;
	if(arg.recursive)
		arg.mp.lookupflags |= DO_OPEN_DIRS | NO_DOTS;
	exit(main_loop(&arg.mp, argv + optind, argc - optind));
}
